src/DssProxyActionsCharter.sol

```diff
@@ -0,0 +1,656 @@
+        bytes32 ilk = GemJoinLike(gemJoin).ilk();
+        (, uint art) = VatLike(vat).urns(ilk, urn);
+
+        // Joins DAI amount into the vat
+        daiJoin_join(daiJoin, _getWipeAllWad(vat, urn, ilk));
+        uint wadC = convertTo18(gemJoin, amtC);
+        // Paybacks debt to the CDP and unlocks token amount from it
+        frob(manager, ilk, usr, -toInt(wadC), -int(art));
+        // Moves the amount from the CDP urn to proxy's address
+        flux(manager, ilk, usr, address(this), wadC);
+        // Exits token amount to the user's wallet as a token
+        ManagerLike(manager).exit(gemJoin, address(this), amtC);
+    }
+}
+
+contract DssProxyActionsEndCharter is Common {
+    // Internal functions
+
+    function _free(
+        address manager,
+        address end,
+        bytes32 ilk,
+        address usr
+    ) internal returns (uint ink) {
+        address urn = ManagerLike(manager).getOrCreateProxy(usr);
+        VatLike vat = VatLike(ManagerLike(manager).vat());
+        uint art;
+        (ink, art) = vat.urns(ilk, urn);
+
+        // If CDP still has debt, it needs to be paid
+        if (art > 0) {
+            EndLike(end).skim(ilk, urn);
+            (ink,) = vat.urns(ilk, urn);
+        }
+        // Approves the manager to transfer the position to proxy's address in the vat
+        if (vat.can(address(this), address(manager)) == 0) {
+            vat.hope(manager);
+        }
+        // Transfers position from CDP to the proxy address
+        ManagerLike(manager).quit(ilk, address(this));
+        // Frees the position and recovers the collateral in the vat registry
+        EndLike(end).free(ilk);
+    }
+
+    // Public functions
+    function freeETH(
+        address manager,
+        address ethJoin,
+        address end,
+        address usr
+    ) public {
+        uint wad = _free(manager, end, GemJoinLike(ethJoin).ilk(), usr);
+        // Exits WETH amount to proxy address as a token
+        ManagerLike(manager).exit(ethJoin, address(this), wad);
+        // Converts WETH to ETH
+        GemJoinLike(ethJoin).gem().withdraw(wad);
+        // Sends ETH back to the user's wallet
+        msg.sender.transfer(wad);
+    }
+
+    function freeGem(
+        address manager,
+        address gemJoin,
+        address end,
+        address usr
+    ) public {
+        uint amt = _free(manager, end, GemJoinLike(gemJoin).ilk(), usr) / 10 ** (18 - GemJoinLike(gemJoin).dec());
+        // Exits token amount to the user's wallet as a token
+        ManagerLike(manager).exit(gemJoin, msg.sender, amt);
+    }
+
+    function pack(
+        address daiJoin,
+        address end,
+        uint wad
+    ) public {
+        daiJoin_join(daiJoin, wad);
+        VatLike vat = DaiJoinLike(daiJoin).vat();
+        // Approves the end to take out DAI from the proxy's balance in the vat
+        if (vat.can(address(this), address(end)) == 0) {
+            vat.hope(end);
+        }
+        EndLike(end).pack(wad);
+    }
```

Comment:
I think since we ended up modifying and leaving all the other relevant functions we can leave `pack` even though it's unchanged. This will ease stuff for integrations as they only need to work with this proxy-actions. If there were many of these untouched functions it might have been best to remove them.

Reply:
Well the thing is `pack` isn't actually part of the integration of these different collaterals.
`pack` is a function for DAI holders after ES which will receive any sort of collateral.
That's where this makes me a bit hesitant as this function has nothing to do with a vault holder wanting to manage an institutional vault or a DAI holder doing a specific cashing of this collateral.
Anyway it is probably find to keep it just for making tests easier.